### PR TITLE
Tests with adam server stub

### PR DESCRIPTION
**Pull request by Malinskiy**

### CODE REVIEW

File: buildSrc/src/main/kotlin/Versions.kt [hunk 1]
```diff
@@ -16,7 +16,7 @@ object Versions {
     val junitGradle = "1.0.0"
     val androidGradleVersion = "4.0.0"
 
-    val junit5 = "5.6.0"
+    val junit5 = "5.7.2"
```
Comment (Malinskiy) [inline buildSrc/src/main/kotlin/Versions.kt:19]: Needed for calling suspend fun inside assertThrows block

File: cli/src/main/kotlin/com/malinskiy/marathon/cli/ApplicationView.kt [hunk 2]
```diff
@@ -52,7 +52,9 @@ fun main(args: Array<String>): Unit = mainBody(
 
             val shouldReportFailure = !configuration.ignoreFailures
             if (!success && shouldReportFailure) {
-                throw SystemExitException("Build failed", 1)
+                throw SystemExitException("Test run failed", 1)
+            } else {
+                throw SystemExitException("Test run finished", 0)
```
Comment (Malinskiy) [inline cli/src/main/kotlin/com/malinskiy/marathon/cli/ApplicationView.kt:57]: This effectively calls System.exit(0)

File: core/src/main/kotlin/com/malinskiy/marathon/io/FileManager.kt [hunk 3]
```diff
@@ -17,6 +17,12 @@ class FileManager(private val output: File) {
         return createFile(directory, filename)
     }
 
+    fun createFile(fileType: FileType, pool: DevicePoolId, device: DeviceInfo, testBatchId: String): File {
```
Comment (Malinskiy) [inline core/src/main/kotlin/com/malinskiy/marathon/io/FileManager.kt:20]: Added new attachments for batch: incomplete batch will save the last screen capture

File: vendor/vendor-android/adam/src/main/kotlin/com/malinskiy/marathon/android/adam/AdamAndroidDevice.kt [hunk 4]
```diff
@@ -123,18 +124,23 @@ class AdamAndroidDevice(
             val local = File(localFilePath)
 
             measureFileTransfer(local) {
-                val channel = client.execute(
-                    CompatPullFileRequest(remoteFilePath, local, supportedFeatures, coroutineScope = this),
-                    serial = adbSerial
-                )
-                for (update in channel) {
-                    progress = update
+                val stat = client.execute(CompatStatFileRequest(remoteFilePath, supportedFeatures), serialNumber)
```
Comment (Malinskiy) [inline vendor/vendor-android/adam/src/main/kotlin/com/malinskiy/marathon/android/adam/AdamAndroidDevice.kt:127]: Saves a lot of time so that we don't pull a file that doesn't exist

File: vendor/vendor-android/adam/src/main/kotlin/com/malinskiy/marathon/android/adam/AdamAndroidDevice.kt [hunk 5]
```diff
@@ -228,7 +234,7 @@ class AdamAndroidDevice(
         }
     }
 
-    override suspend fun installPackage(absolutePath: String, reinstall: Boolean, optionalParams: String): String? {
+    override suspend fun installPackage(absolutePath: String, reinstall: Boolean, optionalParams: List<String>): String? {
```
Comment (Malinskiy) [inline vendor/vendor-android/adam/src/main/kotlin/com/malinskiy/marathon/android/adam/AdamAndroidDevice.kt:237]: Args should come as a list rather than a string with spaces

File: vendor/vendor-android/adam/src/main/kotlin/com/malinskiy/marathon/android/adam/AdamDeviceProvider.kt [hunk 6]
```diff
@@ -155,6 +155,7 @@ class AdamDeviceProvider(
             deviceEventsChannel.cancel()
         }
         logcatManager.close()
+        client.socketFactory.close()
```
Comment (Malinskiy) [inline vendor/vendor-android/adam/src/main/kotlin/com/malinskiy/marathon/android/adam/AdamDeviceProvider.kt:158]: Should stop test runs from freezing: vertx creates daemon threads and we have to close them

File: vendor/vendor-android/base/src/main/kotlin/com/malinskiy/marathon/android/AndroidAppInstaller.kt [hunk 7]
```diff
@@ -65,17 +65,14 @@ class AndroidAppInstaller(configuration: Configuration) {
...
-        } else {
-            "-r"
-        }
-
-        return if (androidConfiguration.installOptions.isNotEmpty()) {
-            "$options ${androidConfiguration.installOptions}"
-        } else {
-            options
-        }
+    private fun optionalParams(device: AndroidDevice): List<String> {
```
Comment (Malinskiy) [inline vendor/vendor-android/base/src/main/kotlin/com/malinskiy/marathon/android/AndroidAppInstaller.kt:68]: Tests surfaced a bug where we effectively send "-r -r -g"

File: vendor/vendor-android/base/src/main/kotlin/com/malinskiy/marathon/android/configuration/TimeoutConfiguration.kt [hunk 8]
```diff
@@ -10,7 +10,7 @@ data class TimeoutConfiguration(
     var pullFile: Duration = Duration.ofSeconds(30),
     var uninstall: Duration = shell,
     var install: Duration = shell,
-    var screenrecorder: Duration = Duration.ofMinutes(10),
+    var screenrecorder: Duration = Duration.ofSeconds(200),
```
Comment (Malinskiy) [inline vendor/vendor-android/base/src/main/kotlin/com/malinskiy/marathon/android/configuration/TimeoutConfiguration.kt:13]: screenrecord supports videos up to 180 seconds anyways

File: vendor/vendor-android/base/src/main/kotlin/com/malinskiy/marathon/android/executor/listeners/screenshot/ScreenCapturerTestRunListener.kt [hunk 9]
```diff
@@ -88,9 +89,13 @@ class ScreenCapturerTestRunListener(
         /**
          * We might not observe the testEnded event, but the testRunFailed will always be reported
          */
-        supervisorJob?.let {
-            logger.debug { "Closing dangling screencapturer job" }
-            it.cancel()
+        supervisorJob?.cancelAndJoin()
+        lastTestIdentifier?.let { id ->
+            val existingRecording = fileManager.createFile(FileType.SCREENSHOT, pool, device.toDeviceInfo(), id.toTest(), testBatchId)
+            if (existingRecording.length() > 0) {
```
Comment (Malinskiy) [inline vendor/vendor-android/base/src/main/kotlin/com/malinskiy/marathon/android/executor/listeners/screenshot/ScreenCapturerTestRunListener.kt:95]: we don't need to check if device is healthy here since the local gif is either created or not

File: vendor/vendor-android/base/src/main/kotlin/com/malinskiy/marathon/android/executor/listeners/video/ScreenRecorderTestRunListener.kt [hunk 10]
```diff
@@ -62,43 +67,53 @@ class ScreenRecorderTestRunListener(
...
         pullVideo(test.toTest())
+        lastRemoteFile = null
     }
 
     override suspend fun testRunFailed(errorMessage: String) {
-        val join = measureTimeMillis {
-            logger.trace { "cancel" }
-            supervisorJob?.cancelAndJoin()
+        try {
+            if (device.verifyHealthy()) {
```
Comment (Malinskiy) [inline vendor/vendor-android/base/src/main/kotlin/com/malinskiy/marathon/android/executor/listeners/video/ScreenRecorderTestRunListener.kt:81]: device might not be available. we have to quickly establish if it's working, otherwise the batch result will be delayed

File: vendor/vendor-android/base/src/main/kotlin/com/malinskiy/marathon/android/executor/listeners/video/ScreenRecorderTestRunListener.kt [hunk 11]
```diff
@@ -62,43 +67,53 @@ class ScreenRecorderTestRunListener(
...
+        } catch (e: InterruptedException) {
+            logger.warn { "Can't stop recording" }
+        } catch (e: TransferException) {
+            logger.warn { "Can't pull video" }
         }
-        logger.trace { "join ${join}ms" }
+    }
+
+    private suspend fun stop() {
         val stop = measureTimeMillis {
```
Comment (Malinskiy) [inline vendor/vendor-android/base/src/main/kotlin/com/malinskiy/marathon/android/executor/listeners/video/ScreenRecorderTestRunListener.kt:96]: stop first, otherwise the video file is corrupted

File: vendor/vendor-android/base/src/main/kotlin/com/malinskiy/marathon/android/model/TestRunResultsAccumulator.kt [hunk 12]
```diff
@@ -159,47 +157,9 @@ class TestRunResultsAccumulator {
...
-            }
-        } else {
-            this.runMetrics.putAll(runMetrics)
-        }
+        this.runMetrics.putAll(runMetrics)
         this.elapsedTime += elapsedTime
         isRunComplete = true
         fillEndTime()
     }
-
```
Comment (Malinskiy) [inline vendor/vendor-android/base/src/main/kotlin/com/malinskiy/marathon/android/model/TestRunResultsAccumulator.kt:175]: This whole piece ported from ddmlib doesn't make any sense

File: vendor/vendor-android/adam/src/main/kotlin/com/malinskiy/marathon/android/adam/AdamAndroidDevice.kt (hunk 4 above)
Comment (sergioboi) [reply vendor/vendor-android/adam/src/main/kotlin/com/malinskiy/marathon/android/adam/AdamAndroidDevice.kt:127]: How does this request actually work? What is the adb command used?

File: vendor/vendor-android/base/src/main/kotlin/com/malinskiy/marathon/android/executor/listeners/screenshot/ScreenCapturerTestRunListener.kt [hunk 13]
```diff
@@ -66,17 +68,16 @@ class ScreenCapturerTestRunListener(
     override suspend fun testEnded(test: TestIdentifier, testMetrics: Map<String, String>) {
         super.testEnded(test, testMetrics)
         val toTest = test.toTest()
+        lastTestIdentifier = null
         logger.debug { "Finished recording for ${toTest.toSimpleSafeTestName()}" }
         supervisorJob?.cancelAndJoin()
         if (!hasFailed && screenRecordingPolicy == ScreenRecordingPolicy.ON_FAILURE) {
             supervisorJob?.invokeOnCompletion {
```
Comment (sergioboi) [inline vendor/vendor-android/base/src/main/kotlin/com/malinskiy/marathon/android/executor/listeners/screenshot/ScreenCapturerTestRunListener.kt:75]: Is `invokeOnCompletion` called when `job` is canceled?

File: vendor/vendor-android/base/src/main/kotlin/com/malinskiy/marathon/android/executor/listeners/video/ScreenRecorderTestRunListener.kt [hunk 14]
```diff
@@ -45,11 +47,14 @@ class ScreenRecorderTestRunListener(
 
     private var hasFailed: Boolean = false
```
Comment (sergioboi) [inline vendor/vendor-android/base/src/main/kotlin/com/malinskiy/marathon/android/executor/listeners/video/ScreenRecorderTestRunListener.kt:48]: Is it okay that we store `hasFailed` status in each listener separatly?

File: vendor/vendor-android/base/src/main/kotlin/com/malinskiy/marathon/android/executor/listeners/video/ScreenRecorderTestRunListener.kt [hunk 15]
```diff
@@ -62,43 +67,53 @@ class ScreenRecorderTestRunListener(
...
+        //Treating as failure
+        hasFailed = true
     }
 
     override suspend fun testEnded(test: TestIdentifier, testMetrics: Map<String, String>) {
         pullVideo(test.toTest())
+        lastRemoteFile = null
     }
 
     override suspend fun testRunFailed(errorMessage: String) {
```
Comment (sergioboi) [inline vendor/vendor-android/base/src/main/kotlin/com/malinskiy/marathon/android/executor/listeners/video/ScreenRecorderTestRunListener.kt:79]: Is this fun called after needed attempts were executed?

File: vendor/vendor-android/base/src/main/kotlin/com/malinskiy/marathon/android/executor/listeners/video/ScreenRecorderTestRunListener.kt [hunk 16]
```diff
@@ -116,11 +131,27 @@ class ScreenRecorderTestRunListener(
...
         val millis = measureTimeMillis {
             device.fileManager.removeRemotePath(remoteFilePath)
         }
         logger.trace { "Removed file in ${millis}ms $remoteFilePath" }
     }
 }
+
+private suspend fun AndroidDevice.verifyHealthy(): Boolean {
+    return withTimeoutOrNull(Duration.ofSeconds(10)) {
+        executeShellCommand("echo quickhealthcheck")?.let { it.contains("quickhealthcheck") } ?: false
```
Comment (sergioboi) [inline vendor/vendor-android/base/src/main/kotlin/com/malinskiy/marathon/android/executor/listeners/video/ScreenRecorderTestRunListener.kt:155]: What about adding timeout as a mandatory argument. Do we have a scenario where we should wait infinite for executing shell command?

File: vendor/vendor-android/adam/src/main/kotlin/com/malinskiy/marathon/android/adam/AdamAndroidDevice.kt (hunk 4 above)
Comment (Malinskiy) [reply vendor/vendor-android/adam/src/main/kotlin/com/malinskiy/marathon/android/adam/AdamAndroidDevice.kt:127]: There is no adb command to issue a stat request. It's an internal adb request

File: vendor/vendor-android/base/src/main/kotlin/com/malinskiy/marathon/android/executor/listeners/screenshot/ScreenCapturerTestRunListener.kt (hunk 13 above)
Comment (Malinskiy) [reply vendor/vendor-android/base/src/main/kotlin/com/malinskiy/marathon/android/executor/listeners/screenshot/ScreenCapturerTestRunListener.kt:75]: yes, precisely

File: vendor/vendor-android/base/src/main/kotlin/com/malinskiy/marathon/android/executor/listeners/video/ScreenRecorderTestRunListener.kt (hunk 14 above)
Comment (Malinskiy) [reply vendor/vendor-android/base/src/main/kotlin/com/malinskiy/marathon/android/executor/listeners/video/ScreenRecorderTestRunListener.kt:48]: Each listener is created per batch. If you want I can rename the variable to hasBatchFailed

File: vendor/vendor-android/base/src/main/kotlin/com/malinskiy/marathon/android/executor/listeners/video/ScreenRecorderTestRunListener.kt (hunk 15 above)
Comment (Malinskiy) [reply vendor/vendor-android/base/src/main/kotlin/com/malinskiy/marathon/android/executor/listeners/video/ScreenRecorderTestRunListener.kt:79]: It depends. The lifecycle of am instrument output sometimes doesn't send testEnded event in edge-cases, but this method is called after each attempt by default regardless of the test result. If no other methods are reported, then the test is considered passed

File: vendor/vendor-android/base/src/main/kotlin/com/malinskiy/marathon/android/executor/listeners/video/ScreenRecorderTestRunListener.kt (hunk 16 above)
Comment (Malinskiy) [reply vendor/vendor-android/base/src/main/kotlin/com/malinskiy/marathon/android/executor/listeners/video/ScreenRecorderTestRunListener.kt:155]: Not for this case. For all the other cases we have a specific timeout to do that. In this case we're not prepared to wait a long time till the shell responds (20 seconds by default). We just want to quickly establish if the device is healthy or not and pull the file or don't even try to pull it.

File: vendor/vendor-android/base/src/main/kotlin/com/malinskiy/marathon/android/executor/listeners/video/ScreenRecorderTestRunListener.kt (hunk 14 above)
Comment (sergioboi) [reply vendor/vendor-android/base/src/main/kotlin/com/malinskiy/marathon/android/executor/listeners/video/ScreenRecorderTestRunListener.kt:48]: It will be clearer or could rename `TestRunListener` to `TestBatchListener`

Comment (Malinskiy) [reply vendor/vendor-android/base/src/main/kotlin/com/malinskiy/marathon/android/executor/listeners/video/ScreenRecorderTestRunListener.kt:48]: Done

**Comment by Malinskiy** [comment]
Fixes #462

**Review by tagantroy — APPROVED** [review]
